### PR TITLE
solang 0.1.12 (new formula)

### PR DESCRIPTION
**Pull request by LucasSte**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This PR introduces a formula for Solang, an open-source Solidity compiler. Solang is a front-end for LLVM and we statically link it with LLVM libraries.

To build it from source, we need Solana's LLVM fork, whose changes are not on LLVM official repository. To bypass this, we clone the code and compile it.

### CODE REVIEW

**Comment by SMillerDev** [comment]
> and we statically link it with LLVM libraries.

That's not really acceptable for package manager usage.

> To build it from source, we need Solana's LLVM fork, whose changes are not on LLVM official repository. To bypass this, we clone the code and compile it.

That goes against https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

**Comment by seanyoung** [comment]
> > and we statically link it with LLVM libraries.
> 
> That's not really acceptable for package manager usage.
> 
> > To build it from source, we need Solana's LLVM fork, whose changes are not on LLVM official repository. To bypass this, we clone the code and compile it.
> 
> That goes against https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Totally, and we are unhappy with this too. Solana uses the linux kernel BPF virtual machine, but in a modified form. We've tried to upstream our changes as a sub-target of the LLVM BPF target. This was not acceptable to the LLVM maintainers, so we are a creating an entirely new target within LLVM (independant of BPF). This is a substantial amount of work, which is under way but not ready for review yet (this coincides with some architectural changes too). Getting this ready for review, reviewed, merged and in an llvm release is some time away.

Maintaining an LLVM fork is a real pain both for us and our users. We look forward to the day when this is over, absolutely. However, until we get to this point, how can we get the compiler into brew? An exception or something like that?

**Comment by carlocab** [comment]
> A quick grep for vendored in the `Formula/` dir shows many vendored libraries, so it looks like exceptions are made in some circumstances.

We're not happy about those either. In any case, there are many formulae in this repo, and some violating our policies isn't really a good reason to add more formulae that do the same.

Do you have a timeline for when the changes needed will be upstreamed?

**Comment by seanyoung** [comment]
> > A quick grep for vendored in the `Formula/` dir shows many vendored libraries, so it looks like exceptions are made in some circumstances.
> 
> We're not happy about those either. In any case, there are many formulae in this repo, and some violating our policies isn't really a good reason to add more formulae that do the same.
> 
> Do you have a timeline for when the changes needed will be upstreamed?

There is no timeline for this work yet. Creating a new llvm target is a multi year effort, it really is a large piece of work.

Note that even with the target upstreamed into llvm, there is no guarantee that we will be able to use the upstreamed version. We may need changes faster than that llvm can upstream them. This is quite common for a compiler: rust and swift carry their own vendored version of llvm. The only reason clang does not is because it lives in the same repo as upstream llvm.

I totally agree this is awful from a packaging perspective, but it is the reality we live in.

We could also provide a binary-only cask, would be acceptable?

**Review by carlocab — COMMENTED** [review]
> Note that even with the target upstreamed into llvm, there is no guarantee that we will be able to use the upstreamed version. We may need changes faster than that llvm can upstream them. This is quite common for a compiler: rust and swift carry their own vendored version of llvm. The only reason clang does not is because it lives in the same repo as upstream llvm.

I agree that it is common, but that's different from being acceptable, and other projects probably shouldn't be following the example of Rust and Swift here.

It's probably also worth considering that Rust and Swift are widely used enough that they can get away with things that most projects don't.

That said -- Rust actually does work pretty well on tagged releases of upstream LLVM (barring a handful of obscure/subtle bugs), so I'm not even sure using Rust is a good example here.

> We could also provide a binary-only cask, would be acceptable?

Is this software packaged by any other distros/package managers?

At the moment I'm inclined to think that this formula may be better suited for an external tap for now, and then we can gauge the level of interest users have in it.

File: Formula/solang.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	        }
+
+      	  function get() public view returns (bool) {
+		        return value;
+	        }
+        }
+    EOS
+    system "#{bin}/solang", "--target", "solana", "flipper.sol"
+    assert File.file?("#{testpath}/flipper.abi")
+    assert File.file?("#{testpath}/bundle.so")
```
Comment (carlocab) [inline Formula/solang.rb:46]:
```suggestion
    system bin/"solang", "--target", "solana", "flipper.sol"
    assert_predicate testpath/"flipper.abi", :file?
    assert_predicate testpath/"bundle.so", :file?
```

File: Formula/solang.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  depends_on "rust" => :build
+
+  resource "llvm" do
+    url "https://github.com/solana-labs/llvm-project.git",
+        revision: "920bf6a3c2476f640448cc534399ad17820f10d7"
+  end
+
+  def install
+    resource("llvm").stage do
+      system "cmake", "-S", "llvm-project", "-B", "build",
```
Comment (carlocab) [inline Formula/solang.rb:19]:
```suggestion
      system "cmake", "-S", "llvm", "-B", "build",
```
Oops, sorry

**Review by chenrui333 — APPROVED** [review]

**Comment by carlocab** [comment]
@chenrui333 I don't think this is ready to merge given the conversation in the PR. Please do not ignore existing reviews when making your own.

**Comment by LucasSte** [comment]
> > We could also provide a binary-only cask, would be acceptable?
> 
> Is this software packaged by any other distros/package managers?

No. Adding to Brew is going to be our first release in a package manager.

> At the moment I'm inclined to think that this formula may be better suited for an external tap for now, and then we can gauge the level of interest users have in it.

We can try it. Do you have any other suggestion on your side on how we can get Solang on Brew?

**Comment by SMillerDev** [comment]
> We can try it. Do you have any other suggestion on your side on how we can get Solang on Brew?

The external tap is the suggestion, and IMO the only option to get solang in brew.

**Pull request closed without merging**